// testcafe/lib/proxyless/request-pipeline/special-handlers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const injectables_1 = require("../../assets/injectables");
const debug_loggers_1 = require("../../utils/debug-loggers");
const cdp_1 = require("../utils/cdp");
const constants_1 = require("./constants");
const http_status_codes_1 = require("http-status-codes");
const load_assets_1 = __importDefault(require("../../load-assets"));
const string_1 = require("../utils/string");
const internalRequest = {
    condition: (event) => !event.networkId,
    handler: async (event, client) => {
        (0, debug_loggers_1.requestPipelineInternalRequestLogger)('%r', event);
        await client.Fetch.failRequest({
            requestId: event.requestId,
            errorReason: 'Aborted',
        });
    },
};
const serviceRequest = {
    condition: (event, options, serviceRoutes) => {
        const url = event.request.url;
        // NOTE: the service 'Error page' should be proxied.
        if (url === serviceRoutes.errorPage1
            || url === serviceRoutes.errorPage2)
            return false;
        return options.serviceDomains.some(domain => url.startsWith(domain));
    },
    handler: async (event, client) => {
        (0, debug_loggers_1.requestPipelineServiceRequestLogger)('%r', event);
        const { requestId } = event;
        if ((0, cdp_1.isRequest)(event))
            await client.Fetch.continueRequest({ requestId });
        else {
            // Hack: CDP doesn't allow to continue response for requests sent from the reloaded page.
            // Such situation rarely occurs on 'heartbeat' or 'open-file-protocol' requests.
            // We are using the simplest way to fix it - just omit such errors.
            try {
                await client.Fetch.continueResponse({ requestId });
            }
            catch (err) {
                if (err.message === constants_1.INVALID_INTERCEPTED_RESPONSE_ERROR_MSG)
                    return;
                throw err;
            }
        }
    },
};
const defaultFaviconRequest = {
    condition: (event) => {
        const parsedUrl = new URL(event.request.url);
        return parsedUrl.pathname === injectables_1.DEFAULT_FAVICON_PATH;
    },
    handler: async (event, client, options) => {
        (0, debug_loggers_1.requestPipelineLogger)('%r', event);
        if ((0, cdp_1.isRequest)(event))
            await client.Fetch.continueRequest({ requestId: event.requestId });
        else {
            if (event.responseStatusCode === http_status_codes_1.StatusCodes.NOT_FOUND) { // eslint-disable-line no-lonely-if
                const { favIcon } = (0, load_assets_1.default)(options.developmentMode);
                await client.Fetch.fulfillRequest({
                    requestId: event.requestId,
                    responseCode: http_status_codes_1.StatusCodes.OK,
                    responseHeaders: [constants_1.FAVICON_CONTENT_TYPE_HEADER],
                    body: (0, string_1.toBase64String)(favIcon),
                });
            }
            else
                await client.Fetch.continueResponse({ requestId: event.requestId });
        }
    },
};
const SPECIAL_REQUEST_HANDLERS = [
    internalRequest,
    serviceRequest,
    defaultFaviconRequest,
];
function getSpecialRequestHandler(event, options, serviceRoutes) {
    const specialRequestHandler = SPECIAL_REQUEST_HANDLERS.find(h => h.condition(event, options, serviceRoutes));
    return specialRequestHandler ? specialRequestHandler.handler : null;
}
exports.default = getSpecialRequestHandler;
module.exports = exports.default;
